### PR TITLE
Add saveManyOrFail().

### PR DESCRIPTION
**Pull request by dereuromark**
Refs https://github.com/cakephp/cakephp/issues/12957

### CODE REVIEW

File: tests/TestCase/ORM/TableTest.php [hunk 1]
```diff
@@ -2736,6 +2736,70 @@ public function testSaveManyFailedWithException()
...
+            ->first();
+        $this->assertSame('admad', $first->name);
+    }
+
+    /**
+     * Test saveManyOrFail() with failed save
+     *
+     * @return void
+     */
+    public function testSaveManyOrFailFailed()
```
Comment (markstory) [inline tests/TestCase/ORM/TableTest.php]: Would be good to have a test that covers a unique constraint failure. And also a test covering the attribute roll back.

Comment (dereuromark) [reply tests/TestCase/ORM/TableTest.php]: I am happy if someone wants to help out here, adding more tests. So far I copied the tests of saveMany()

File: src/ORM/Table.php [hunk 2]
```diff
@@ -2086,6 +2086,61 @@ public function saveMany(iterable $entities, $options = [])
...
+     * any one of the records fails to save due to failed validation or database
+     * error.
+     *
+     * @param \Cake\Datasource\EntityInterface[]|\Cake\Datasource\ResultSetInterface $entities Entities to save.
+     * @param array|\ArrayAccess $options Options used when calling Table::save() for each entity.
+     * @return \Cake\Datasource\EntityInterface[]|\Cake\Datasource\ResultSetInterface Entities list.
+     * @throws \Exception
+     * @throws \Cake\ORM\Exception\PersistenceFailedException
+     */
+    public function saveManyOrFail(iterable $entities, $options = [])
```
Comment (ADmad) [inline src/ORM/Table.php]: There's lot of code duplication between this method and `saveMany()`, only difference being this method stores reference to entity which failed to save and then throwing exception. So the common code should be consolidated into a helper method.

Comment (dereuromark) [reply src/ORM/Table.php]: I refactored.
We shouldnt use too many exceptions for workflow :) But I hope the current solution is maximum DRY while still allowing both cases.

**Comment by dereuromark** [comment]
The fails are unrelated afaik.

**Comment by markstory** [comment]
Both scrutinizer and stickler have some relevant errors.

**Comment by dereuromark** [comment]
Sry, I must have overlooked them. With the failing appveeyor it also always looks failing somewhere.

**Pull request merged**